jobmaster/dm/ticker/ticker.go

```diff
@@ -0,0 +1,68 @@
+package ticker
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+type Ticker interface {
+	TickImpl(ctx context.Context) error
+}
+
+// DefaultTicker defines the template method to run periodically.
+type DefaultTicker struct {
+	Ticker
+
+	mu            sync.RWMutex
+	lastCheckTime time.Time
+	nextCheckTime time.Time
+
+	normalInterval time.Duration
+	errorInterval  time.Duration
+}
+
+func NewDefaultTicker(normalInterval, errorInterval time.Duration) *DefaultTicker {
+	defaultTicker := &DefaultTicker{
+		normalInterval: normalInterval,
+		errorInterval:  errorInterval,
+	}
+	defaultTicker.SetNextCheckTime(time.Now())
+	return defaultTicker
+}
+
+func (s *DefaultTicker) SetNextCheckTime(t time.Time) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if (t.After(s.lastCheckTime) && t.Before(s.nextCheckTime)) || s.lastCheckTime.Equal(s.nextCheckTime) {
```

Comment:
It seems `t.After(s.lastCheckTime) && t.Before(s.nextCheckTime)` can't be triggered in code path, or is it better to check `s.lastCheckTime.Equal(s.nextCheckTime)` first to reduce one check.